tree/tree/src/TTreeCacheUnzip.cxx

```diff
@@ -685,235 +532,407 @@ void TTreeCacheUnzip::ResetCache()
+
+           fNUnzip++;
+   }
+   else {
+           fUnzipChunks[reqi] = 0;
+           fUnzipLen[reqi] = 0;
+           fUnzipStatus[reqi] = 2;
+   }
 
-            fUnzipStatus  = aUnzipStatus;
-            fUnzipLen  = aUnzipLen;
-            fUnzipChunks = aUnzipChunks;
+   return 0;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// This tbb task class try unzipping a set of baskets, we use a vector to
+/// store the basket indices to be unzipped by this task.
+
+class UnzipTask: public tbb::task {
+
+private:
+   TTreeCacheUnzip* cache;
+   std::vector<Int_t> indices;
+ 
+public:
+   UnzipTask(TTreeCacheUnzip* c, std::vector<Int_t>& ids) {
+      cache = c;
+      indices = ids;
+   }
+
+   tbb::task* execute() {
+      for (auto ii : indices) {
+         Byte_t oldV = 0;
+         Byte_t newV = 1;
+         if(cache->fUnzipStatus[ii].compare_exchange_weak(oldV, newV, std::memory_order_release, std::memory_order_relaxed)) {
+            Int_t locbuffsz = 16384;
+            char *locbuff = new char[16384];
+            Int_t res = cache->UnzipCache(ii, locbuffsz, locbuff);
+            if(res)
+               if (gDebug > 0)
+                  Info("UnzipCache", "Unzipping failed or cache is in learning state");
+         }
+      }
+      return nullptr;
+   }
+};
+
+////////////////////////////////////////////////////////////////////////////////
+/// This tbb task class allocate UnzipTasks to unzip baskets. We spawn a 
+/// UnzipTask to unzip a set of basksets where their accumulated size goes 
+/// beyond 100KB.
+
+class MappingTask: public tbb::task {
+
+private:
+   TTreeCacheUnzip* cache;
+   std::vector<Int_t> indices;
+
+public:
+   MappingTask(TTreeCacheUnzip* c) {
+      cache = c;
+      indices.clear();
+   }
 
-            fNseekMax  = fNseek;
+   tbb::task* execute() {
+      Int_t accusz = 0;
+      Int_t taskcnt = 0;
+      tbb::task_list tl;
+
+      UnzipTask* t = nullptr;
+      for (Int_t ii = 0; ii < cache->fNseek; ii++) {
+         while (accusz < 102400) {
+            accusz += cache->fSeekLen[ii];
+            indices.push_back(ii);
+            ii++;
+            if (ii >= cache->fNseek) break;
          }
+         if (ii < cache->fNseek) ii--;
+         t = new(this->allocate_child()) UnzipTask(cache, indices);
+         tl.push_back(*t);
+         taskcnt++;
+         indices.clear();
+         accusz = 0;
+      }
+      this->set_ref_count(taskcnt+1);
+      this->spawn(tl);
+      return nullptr;
+   }
+};
+
+////////////////////////////////////////////////////////////////////////////////
+/// We create a dummy task and spawn a MappingTask which spawn actual unzipping
+/// tasks at background without competing with main thread.
+
+Int_t TTreeCacheUnzip::CreateTasks()
+{
+   root = new(tbb::task::allocate_root()) tbb::empty_task;
+
+   root->set_ref_count(2);
+   MappingTask* t = new(root->allocate_child()) MappingTask(this);
+   root->spawn(*t);
```

Comment:
The first point is not provided, but you can create a separate TBB task that runs TThreadExecutor too :)

On the second point... you can call TThreadExecutor::Map and get back the vector of status arrays to save them after the Map call. Don't know if this fits your particular case.